### PR TITLE
build(fluid-runner): use node16 module resolution 

### PR DESCRIPTION
**Pull request by sonalideshpandemsft**
Additions:
1. Update `parseBundleAndExportFile.ts` to use import 
```ts
const codeLoaderBundle = await import(codeLoader);
```
2. Upgrade `yargs` to `17.7.2` and `@types/yargs` to `17.0.32`. As of v16, yargs supports ESM imports
3. Add sample executable scripts in both ESM and CJS formats
4. Add tsconfig.bin.lint.json file to extend eslint coverage to the bin folder

Original PR: https://github.com/microsoft/FluidFramework/pull/20245

### CODE REVIEW

File: packages/tools/fluid-runner/src/index.ts [hunk 1]
```diff
@@ -4,15 +4,21 @@
...
+export { ICodeLoaderBundle, IFluidFileConverter } from "./codeLoaderBundle.js";
+export {
+	createContainerAndExecute,
+	exportFile,
+	IExportFileResponse,
+	IExportFileResponseSuccess,
+	IExportFileResponseFailure,
+} from "./exportFile.js";
+export { fluidRunner } from "./fluidRunner.js";
+export { OutputFormat, ITelemetryOptions, IFileLogger } from "./logger/fileLogger.js";
```
Comment (jason-ha) [inline packages/tools/fluid-runner/src/index.ts:16]: I see exports were added to resolve api-extractor issues. Do we know the history for the prior choices? ...to go against api-extractor and not export those? Such changes are best done explicitly and separately.

**Review by jason-ha — CHANGES_REQUESTED** [review]
Description should be updated to note the ESM support.

File: packages/tools/fluid-runner/bin/fluid-runner.cjs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env node
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+require("../dist/fluidRunner.js").fluidRunner();
```
[the remark this answers is not in the document]
Comment (sonalideshpandemsft) [reply packages/tools/fluid-runner/bin/fluid-runner.cjs]: Converting this to ESM requires some update to fluidRunner. As of v16, yargs supports ESM imports

File: packages/tools/fluid-runner/src/test/fluidRunner.spec.ts [hunk 3]
```diff
@@ -192,11 +193,11 @@ describe("fluid-runner from command line", () => {
 });
 
 describe("custom fluidFileConverter provided", () => {
-	const command = path.join(__dirname, "../../src/test/sampleCodeLoaders", "sample-executable");
+	const command = path.join(_dirname, "../../bin/sample-executable.cjs");
```
[the remark this answers is not in the document]
Comment (sonalideshpandemsft) [reply packages/tools/fluid-runner/src/test/fluidRunner.spec.ts]: The move was made just to have all scripts under bin (organizational purpose only)

File: packages/tools/fluid-runner/bin/fluid-runner.mjs
Comment (jason-ha) [inline packages/tools/fluid-runner/bin/fluid-runner.mjs:1]: Does .mjs extension by us something? Or was there a benefit to having no extension before?
From what I know .mjs is useful when it needs interpreted as ESM, but this package has "type": "module"; so .js or no extension should already be expected to be ESM. (I tend to assume someone had a reason for no extension, but I don't know what it is.) I do like the explicitness of .mjs; so, no need to change.

File: packages/tools/fluid-runner/src/test/sampleCodeLoaders/sample-executable.mjs
Comment (jason-ha) [inline packages/tools/fluid-runner/src/test/sampleCodeLoaders/sample-executable.mjs:1]: This means missing testing of this for CJS, which we should have if we are keeping CJS support. I think I have a kludge for how to test both.

**Review by jason-ha — CHANGES_REQUESTED** [review]
I'll try to push some lint and correctness fixes Tuesday.

File: packages/tools/fluid-runner/bin/fluid-runner.mjs
Comment (sonalideshpandemsft) [reply packages/tools/fluid-runner/bin/fluid-runner.mjs:1]: afair, mocha test failures on the old PR suggested to add extension (js/mjs/cjs) to this script: https://dev.azure.com/fluidframework/public/_build/results?buildId=250799&view=logs&j=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3&t=5cfe65a4-d7cb-5934-e953-1c67617821f4

```ts
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension "" for /mnt/vss/_work/1/s/packages/tools/fluid-runner/bin/fluid-runner. Loading extensionless files is not supported inside of "type":"module" package.json contexts. The package.json file /mnt/vss/_work/1/s/packages/tools/fluid-runner/package.json caused this "type":"module" context. Try changing /mnt/vss/_work/1/s/packages/tools/fluid-runner/bin/fluid-runner to have a file extension. Note the "bin" field of package.json can point to a file with an extension, for example {"type":"module","bin":{"fluid-runner":"./bin/fluid-runner.js"}}
packages/tools/fluid-runner test:mocha:     at new NodeError (node:internal/errors:405:5)
packages/tools/fluid-runner test:mocha:     at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:79:11)
packages/tools/fluid-runner test:mocha:     at defaultGetFormat (node:internal/modules/esm/get_format:124:36)
packages/tools/fluid-runner test:mocha:     at defaultLoad (node:internal/modules/esm/load:84:20)
packages/tools/fluid-runner test:mocha:     at nextLoad (node:internal/modules/esm/loader:163:28)
packages/tools/fluid-runner test:mocha:     at ESMLoader.load (node:internal/modules/esm/loader:603:26)
packages/tools/fluid-runner test:mocha:     at ESMLoader.moduleProvider (node:internal/modules/esm/loader:457:22)
packages/tools/fluid-runner test:mocha:     at new ModuleJob (node:internal/modules/esm/module_job:64:26)
packages/tools/fluid-runner test:mocha:     at #createModuleJob (node:internal/modules/esm/loader:480:17)
packages/tools/fluid-runner test:mocha:     at ESMLoader.getModuleJob (node:internal/modules/esm/loader:434:34) {
packages/tools/fluid-runner test:mocha:   code: 'ERR_UNKNOWN_FILE_EXTENSION'
```

File: packages/tools/fluid-runner/tsconfig.bin.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+	"extends": "../../../common/build/build-common/tsconfig.node16.json",
+	"include": ["bin/*"],
+	"references": [
```
Comment (sonalideshpandemsft) [inline packages/tools/fluid-runner/tsconfig.bin.json:4]: why is tsconfig.bin.json required?

File: packages/tools/fluid-runner/src/test/sampleCodeLoaders/sample-executable.cjs.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env node
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { fluidRunner } from "../../index.js";
```
Comment (sonalideshpandemsft) [inline packages/tools/fluid-runner/src/test/sampleCodeLoaders/sample-executable.cjs.ts:7]: why import differently? why not the below? 
```ts
import { fluidRunner } from "@fluidframework/fluid-runner";
```

File: packages/tools/fluid-runner/src/test/fluidRunner.spec.ts [hunk 6]
```diff
@@ -193,11 +193,13 @@ describe("fluid-runner from command line", () => {
...
-describe("custom fluidFileConverter provided", () => {
+describe(`custom fluidFileConverter provided (${
+	_dirname.includes("dist") ? "CJS" : "ESM"
+})`, () => {
 	const command = path.join(
 		_dirname,
-		"../../src/test/sampleCodeLoaders",
-		"sample-executable.mjs",
+		"sampleCodeLoaders",
+		_dirname.includes("dist") ? "sample-executable.cjs.js" : "sample-executable.esm.js",
```
Comment (sonalideshpandemsft) [inline packages/tools/fluid-runner/src/test/fluidRunner.spec.ts:202]: love this change. let's us test our script with both ESM and CJS :)

File: packages/tools/fluid-runner/tsconfig.bin.json (hunk 4 above)
Comment (jason-ha) [reply packages/tools/fluid-runner/tsconfig.bin.json:4]: We need some TS project file to provide coverage with eslint. Probably should rename it to have .`lint` in the name like was done for CJS test case.

File: packages/tools/fluid-runner/src/index.ts (hunk 1 above)
Comment (sonalideshpandemsft) [reply packages/tools/fluid-runner/src/index.ts:16]: api-extractor forgotten-export warning were suppressed before: https://github.com/microsoft/FluidFramework/pull/20349/files#diff-04ef5f87a3b831c02ccb64223b9bfc0bb33d0403d8b2f250179dbaf9e50267f3

File: packages/tools/fluid-runner/src/test/sampleCodeLoaders/sample-executable.cjs.ts (hunk 5 above)
Comment (jason-ha) [reply packages/tools/fluid-runner/src/test/sampleCodeLoaders/sample-executable.cjs.ts:7]: "@fluidframework/fluid-runner" is the nice way to do it. You could take this example and use it outside of this package.

In the CJS case (surprise to me) the "self-reference" was an error according to VS Code's tsc result. I didn't think that support mattered CJS/ESM and just confirmed it looks like a VS Code limitation/bug. tsc is happy with @fluidframework/fluid-runner. I will update. Thanks for asking. :)

**Review by jason-ha — APPROVED** [review]

File: packages/tools/fluid-runner/src/test/sampleCodeLoaders/sample-executable.cjs.ts (hunk 5 above)
Comment (jason-ha) [reply packages/tools/fluid-runner/src/test/sampleCodeLoaders/sample-executable.cjs.ts:11]: Well, I don't know how VS Code is magically right about what should be done. Comment added.

**Pull request merged**